File: dofbot_pro_ws/src/arm_moveit_demo/scripts/03_set_joint_plan.py
#!/usr/bin/env python
# coding: utf-8
import rospy
from math import pi
from time import sleep
from moveit_commander.move_group import MoveGroupCommander

# 角度转弧度
DE2RA = pi / 180

if __name__ == '__main__':
    # 初始化节点
    rospy.init_node("set_joint_py", anonymous=True)
    # 初始化机械臂
    dofbot_pro = MoveGroupCommander("arm_group")
    # 当运动规划失败后，允许重新规划
    dofbot_pro.allow_replanning(True)
    dofbot_pro.set_planning_time(5)
    # 尝试规划的次数
    dofbot_pro.set_num_planning_attempts(10)
    # 设置允许目标角度误差
    dofbot_pro.set_goal_joint_tolerance(0.001)
    # 设置允许的最大速度和加速度
    dofbot_pro.set_max_velocity_scaling_factor(1.0)
    dofbot_pro.set_max_acceleration_scaling_factor(1.0)
    # 设置"down"为目标点
    dofbot_pro.set_named_target("down")
    dofbot_pro.go()
    sleep(0.5)
    # 设置目标点 弧度
    joints = [0.79, 0.79, -1.57, -1.57, 0]
    dofbot_pro.set_joint_value_target(joints)
    # 多次执行,提高成功率
    for i in range(5):
        # 运动规划
        plan = dofbot_pro.plan()
        #print("plan: ",plan)
        if len(plan.joint_trajectory.points) != 0:
            print ("plan success")
            # 规划成功后运行
            dofbot_pro.execute(plan)
            break
        else:
            print ("plan error")

File: dofbot_pro_ws/src/arm_moveit_demo/src/03_set_joint_plan.cpp
#include <iostream>
#include "ros/ros.h"
#include <tf/LinearMath/Quaternion.h>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit_visual_tools/moveit_visual_tools.h>

using namespace std;

int main(int argc, char **argv) {
    ros::init(argc, argv, "set_joint_plan_cpp");
    ros::NodeHandle n;
    ros::AsyncSpinner spinner(1);
    spinner.start();
    moveit::planning_interface::MoveGroupInterface dofbot_pro("arm_group");
    dofbot_pro.allowReplanning(true);
    // 规划的时间(单位：秒)
    dofbot_pro.setPlanningTime(5);
    dofbot_pro.setNumPlanningAttempts(10);
    // 设置允许目标角度误差
    dofbot_pro.setGoalJointTolerance(0.01);
    // 设置允许的最大速度和加速度
    dofbot_pro.setMaxVelocityScalingFactor(1.0);
    dofbot_pro.setMaxAccelerationScalingFactor(1.0);
    dofbot_pro.setNamedTarget("up");
    dofbot_pro.move();
//    sleep(0.1);
    //设置具体位置
    vector<double> pose{0, 0.79, -1.57, -1.57, 0};
    dofbot_pro.setJointValueTarget(pose);
    moveit::planning_interface::MoveGroupInterface::Plan plan;
    const moveit::planning_interface::MoveItErrorCode &code = dofbot_pro.plan(plan);
    if (code == code.SUCCESS) {
        ROS_INFO_STREAM("plan success");
        // 显示轨迹
        string frame = dofbot_pro.getPlanningFrame();
        moveit_visual_tools::MoveItVisualTools tool(frame);
        tool.deleteAllMarkers();
        tool.publishTrajectoryLine(plan.trajectory_, dofbot_pro.getCurrentState()->getJointModelGroup("arm_group"));
        tool.trigger();
        dofbot_pro.execute(plan);
    } else {
        ROS_INFO_STREAM("plan error");
    }
    return 0;
}

